### PR TITLE
Feature/create edit

### PR DESCRIPTION
**Pull request by Yassyass181104**

### CODE REVIEW

File: app/src/main/java/com/android/sample/ui/dialogs/AddImageDialog.kt
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/dialogs/AddImageDialog.kt:1]: Only reformatting, seems good.

File: app/src/main/java/com/android/sample/ui/dialogs/AddUserDialog.kt
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/dialogs/AddUserDialog.kt:1]: Good that you switched it to a dialog. How about a cancel button ?

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt [hunk 1]
```diff
@@ -35,32 +39,44 @@ import androidx.compose.ui.Alignment
...
-  var location by remember { mutableStateOf(activity?.location) }
-  var price by remember { mutableStateOf(activity?.price.toString()) }
-  var placesLeft by remember { mutableStateOf(activity?.placesLeft.toString()) }
+  var title by remember { mutableStateOf(activity?.title ?: "") }
+  var description by remember { mutableStateOf(activity?.description ?: "") }
+  val creator by remember { mutableStateOf(activity?.creator ?: "") }
+  var location by remember { mutableStateOf(activity?.location ?: "") }
+  var price by remember { mutableStateOf(activity?.price.toString() ?: "") }
+  var placesLeft by remember { mutableStateOf(activity?.placesLeft.toString() ?: "") }
+  var attendees by remember { mutableStateOf(activity?.participants ?: listOf()) }
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:79]: Good that you added support for null cases.

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt [hunk 2]
```diff
@@ -96,9 +112,10 @@ fun EditActivityScreen(
       }) { paddingValues ->
         Column(
             modifier =
-                Modifier.padding(paddingValues).fillMaxSize().background(color = Color(0xFFFFFFFF))
-                    .verticalScroll(rememberScrollState())
-                    ,
+                Modifier.padding(paddingValues)
+                    .fillMaxSize()
+                    .background(color = Color(0xFFFFFFFF))
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:117]: You should use theme colors or at least predefined constants for the color.

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt [hunk 3]
```diff
@@ -151,9 +168,75 @@ fun EditActivityScreen(
...
+                        contentDescription = "add a new attendee",
+                    )
+                    Text("Add Attendee")
+                  }
+                }
+
+                LazyRow(
+                    modifier = Modifier.fillMaxHeight().height(85.dp).padding(8.dp),
+                ) {
+                  items(attendees.size) { index ->
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:194]: we should add the definition of what is an "attendee" in the wiki (in the conventions document)

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt [hunk 4]
```diff
@@ -151,9 +168,75 @@ fun EditActivityScreen(
...
+                onDismiss = { showDialog = false },
+                onAddUser = { user -> attendees = attendees + user })
+          }
           Spacer(modifier = Modifier.height(32.dp))
           Button(
-              enabled = title!!.isNotEmpty() && description!!.isNotEmpty() && dueDate.isNotEmpty(),
+              enabled =
+                  title?.isNotEmpty() ?: false &&
+                      description?.isNotEmpty() ?: false &&
+                      dueDate.isNotEmpty(),
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:239]: try using this : 
```
enabled = title.orEmpty().isNotEmpty() &&
          description.orEmpty().isNotEmpty() &&
          dueDate.isNotEmpty()
```


it manages the null cases with a better readability

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt [hunk 5]
```diff
@@ -226,9 +314,32 @@ fun EditActivityScreen(
                   Icons.Outlined.Delete,
                   contentDescription = "add a new activity",
               )
-              Text("Create")
+              Text("Delete")
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:317]: very important fix ! I have had an issue with this in the previous version, where I deleted an activity instead of creating it

File: app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt [hunk 6]
```diff
@@ -61,149 +62,217 @@ fun CreateActivityScreen(
...
-                    contentDescription = "Back")
-              }
-            })
+        )
+      },
+      bottomBar = {
+        BottomNavigationMenu(
+            onTabSelect = { route -> navigationActions.navigateTo(route) },
+            tabList = LIST_TOP_LEVEL_DESTINATION,
+            selectedItem = Route.ADD_ACTIVITY)
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt:90]: good that you added the bottom bar. I guess it was also added in another PR, but since it is the same, we should not have merge issues.

File: app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt [hunk 7]
```diff
@@ -61,149 +62,217 @@ fun CreateActivityScreen(
...
-              value = dueDate,
-              onValueChange = { dueDate = it },
-              label = { Text("Date") },
-              modifier = Modifier.padding(8.dp).fillMaxWidth().testTag("inputDateCreate"),
-              placeholder = { Text("dd/mm/yyyy") },
-          )
-          Spacer(modifier = Modifier.height(8.dp))
+                Modifier.padding(paddingValues)
+                    .fillMaxSize()
+                    .background(color = Color(0xFFFFFFFF))
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt:97]: should use a theme color

File: app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt [hunk 8]
```diff
@@ -61,149 +62,217 @@ fun CreateActivityScreen(
...
+                    .verticalScroll(rememberScrollState())) {
+              Carousel()
+              Spacer(modifier = Modifier.height(8.dp))
+              OutlinedTextField(
+                  value = title,
+                  onValueChange = { title = it },
+                  label = { Text("Title") },
+                  modifier = Modifier.padding(8.dp).fillMaxWidth().testTag("inputTitleCreate"),
+                  placeholder = { Text("Give a title of the activity") },
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt:106]: we should add those strings in the xml constants file

File: app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt [hunk 9]
```diff
@@ -61,149 +62,217 @@ fun CreateActivityScreen(
...
+              )
+              Spacer(modifier = Modifier.height(8.dp))
+              OutlinedTextField(
+                  value = description,
+                  onValueChange = { description = it },
+                  label = { Text("Description") },
+                  modifier =
+                      Modifier.padding(8.dp).fillMaxWidth().testTag("inputDescriptionCreate"),
+                  placeholder = { Text("Describe the activity") },
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt:115]: we should add those strings in the xml constants file

File: app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt [hunk 10]
```diff
@@ -61,149 +62,217 @@ fun CreateActivityScreen(
...
-                            location = location,
-                            images = carouselItems.map{it},
-                            participants = listOf()
+              Spacer(modifier = Modifier.height(8.dp))
+              OutlinedTextField(
+                  value = placesLeft,
+                  onValueChange = { placesLeft = it },
+                  label = { Text("Places Left") },
+                  modifier = Modifier.padding(8.dp).fillMaxWidth().testTag("inputPlacesCreate"),
+                  placeholder = { Text("Places left/Total places") },
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt:141]: we should add those strings in the xml constants file

File: app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt [hunk 11]
```diff
@@ -61,149 +62,217 @@ fun CreateActivityScreen(
...
+              )
+              Spacer(modifier = Modifier.height(8.dp))
+              OutlinedTextField(
+                  value = location,
+                  onValueChange = { location = it },
+                  label = { Text("Location") },
+                  modifier = Modifier.padding(8.dp).fillMaxWidth().testTag("inputLocationCreate"),
+                  placeholder = { Text("Where is it taking place") },
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt:149]: we should add those strings in the xml constants file

File: app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt [hunk 12]
```diff
@@ -61,149 +62,217 @@ fun CreateActivityScreen(
...
+                        Text("Add Attendee")
+                      }
+                    }
+
+                    LazyRow(
+                        modifier = Modifier.fillMaxHeight().height(85.dp).padding(8.dp),
+                    ) {
+                      items(attendees.size) { index ->
+                        Row(
+                            modifier = Modifier.padding(8.dp).background(Color(0xFFFFFFFF)),
```
Comment (hmaryam84) [inline app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt:178]: same remark for theme color

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt (hunk 4 above)
Comment (hmaryam84) [reply app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:239]: but why would this be null ? don't you already manage this when creating the variable ?

File: app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt (hunk 11 above)
Comment (hmaryam84) [reply app/src/main/java/com/android/sample/ui/activity/CreateActivity.kt:149]: we need to have a uniform way of writing placeholders. here you ask a question, but in other equivalent placeholders, we rather give an instruction "Give the title" instead of asking "What is the title?"

that is a good reason to have all the strings in an xml file, to better see and edit this aspect of uniformity

File: app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt [hunk 13]
```diff
@@ -1,99 +1,197 @@
...
+            style = TextStyle(fontWeight = FontWeight.Bold, fontSize = 12.sp),
+        )
+        Button(
+            onClick = { /* no-op */},
+            modifier = Modifier.width(40.dp).height(40.dp).testTag("removeAttendeeButton"),
+        ) {
+          Icon(
+              Icons.Filled.PersonRemove,
+              contentDescription = "remove attendee",
+          )
```
Comment (hmaryam84) [inline app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt:117]: Why do you rewrite the code here explicitly instead of calling the original code ? Doesn't it go against the purpose of testing ?

File: app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt [hunk 14]
```diff
@@ -1,99 +1,197 @@
...
+            style = TextStyle(fontWeight = FontWeight.Bold, fontSize = 12.sp),
+        )
+        Button(
+            onClick = { attendees = attendees.filter { it != attendees[0] } },
+            modifier = Modifier.width(40.dp).height(40.dp).testTag("removeAttendeeButton"),
+        ) {
+          Icon(
+              Icons.Filled.PersonRemove,
+              contentDescription = "remove attendee",
+          )
```
Comment (hmaryam84) [inline app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt:159]: Again, I don't understand why the code is written this way. Where do you call the code to test ?

File: app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt [hunk 15]
```diff
@@ -1,99 +1,197 @@
...
+  }
 
-}
\ No newline at end of file
+  @Test
+  fun parseFraction_returnsNullForEmptyString() {
+    val result = parseFraction("", 0)
+    assertEquals(null, result)
+  }
+}
```
Comment (hmaryam84) [inline app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt:197]: good tests for the parse function. nice that you tested those edge cases.

File: app/src/androidTest/java/com/android/sample/ui/activity/EditActivityScreenTest.kt [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    composeTestRule.onNodeWithTag("inputStatusEdit").performClick()
+    composeTestRule.onNodeWithTag("inputStatusEdit").assertTextContains("Started")
+
+    composeTestRule.onNodeWithTag("inputStatusEdit").performClick()
+    composeTestRule.onNodeWithTag("inputStatusEdit").assertTextContains("Ended")
+
+    composeTestRule.onNodeWithTag("inputStatusEdit").performClick()
+    composeTestRule.onNodeWithTag("inputStatusEdit").assertTextContains("Archived")
+  }
+}*/
```
Comment (hmaryam84) [inline app/src/androidTest/java/com/android/sample/ui/activity/EditActivityScreenTest.kt:105]: all is commented. Is it an expected thing ? I'd recommend to simply remove the file, and keep it locally until needed. If not expected, make sure to uncomment them.

File: app/src/androidTest/java/com/android/sample/ui/dialogs/AddImageDialogTest.kt [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+import org.junit.Test
+
+class AddImageDialogTest {
+
+  @get:Rule val composeTestRule = createComposeRule()
+
+  @Test
+  fun addImageDialogDisplaysCorrectly() {
+    composeTestRule.setContent {
+      AddImageDialog(onDismiss = {}, onGalleryClick = {}, onCameraClick = {})
```
Comment (hmaryam84) [inline app/src/androidTest/java/com/android/sample/ui/dialogs/AddImageDialogTest.kt:16]: Great. Here, instead of the other test file, you really call the code function, which is useful for testing.

File: app/src/androidTest/java/com/android/sample/ui/dialogs/AddImageDialogTest.kt [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    composeTestRule.setContent {
+      AddImageDialog(onDismiss = {}, onGalleryClick = { galleryClicked = true }, onCameraClick = {})
+    }
+
+    composeTestRule.onNodeWithText("Choose from gallery").performClick()
+    assert(galleryClicked)
+  }
+
+  @Test
+  fun addImageDialogCallsOnCameraClick() {
```
Comment (hmaryam84) [inline app/src/androidTest/java/com/android/sample/ui/dialogs/AddImageDialogTest.kt:36]: What happens with this, as we chose not to implement the camera and images ?

File: app/src/androidTest/java/com/android/sample/ui/dialogs/AddUserDialogTest.kt [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+  fun addUserDialogAddsUserOnButtonClick() {
+    var userAdded: SimpleUser? = null
+    composeTestRule.setContent { AddUserDialog(onDismiss = {}, onAddUser = { userAdded = it }) }
+
+    composeTestRule.onNodeWithTag("nameTextFieldUser").performTextInput("John")
+    composeTestRule.onNodeWithTag("surnameTextFieldUser").performTextInput("Doe")
+    composeTestRule.onNodeWithTag("ageTextFieldUser").performTextInput("30")
+    composeTestRule.onNodeWithTag("addUserButton").performClick()
+
+    assert(userAdded == SimpleUser("John", "Doe", 30))
```
Comment (hmaryam84) [inline app/src/androidTest/java/com/android/sample/ui/dialogs/AddUserDialogTest.kt:34]: great test ! maybe you could have used a variable for the simple user fields.

File: app/src/androidTest/java/com/android/sample/ui/dialogs/AddUserDialogTest.kt [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  @Test
+  fun addUserDialogHandlesEmptyFields() {
+    var userAdded: SimpleUser? = null
+    composeTestRule.setContent { AddUserDialog(onDismiss = {}, onAddUser = { userAdded = it }) }
+
+    composeTestRule.onNodeWithTag("addUserButton").performClick()
+
+    assert(userAdded == SimpleUser("", "", 0))
```
Comment (hmaryam84) [inline app/src/androidTest/java/com/android/sample/ui/dialogs/AddUserDialogTest.kt:54]: is it expected to add a user in that case ??

**Review by hmaryam84 — COMMENTED** [review]
Good, in terms of functionnality, it is great. But in terms of refactoring, there are some points to consider. I'd be happy to discuss the points of my comments.

File: app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt [hunk 21]
```diff
@@ -1,99 +1,197 @@
...
+  }
 
-    }
+  @Test
+  fun createActivityScreen_displaysPlacesLeftField() {
+    composeTestRule.setContent { CreateActivityScreen(mockViewModel, mockNavigationActions) }
+    composeTestRule.onNodeWithTag("inputPlacesCreate").assertExists()
+    composeTestRule.onNodeWithTag("inputPlacesCreate").assertIsDisplayed()
+  }
+
```
Comment (DAFERMOH) [inline app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt:76]: Testing if the ui is displayed, LGTM!

File: app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt [hunk 22]
```diff
@@ -1,99 +1,197 @@
...
 
-    @Test
-    fun createActivityScreen_displaysPriceField() {
-        composeTestRule.setContent {
-            CreateActivityScreen(mockViewModel, mockNavigationActions)
+    composeTestRule.setContent {
+      Row(
+          modifier = Modifier.padding(8.dp).background(Color(0xFFFFFFFF)),
+      ) {
+        Text(
```
Comment (DAFERMOH) [inline app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt:93]: I do not get why you are coding some UI here given the signature of the function. You could edit this part or remove it.

File: app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt [hunk 23]
```diff
@@ -1,99 +1,197 @@
...
+            text = attendees[0].name,
+            modifier = Modifier.padding(8.dp),
+            style = TextStyle(fontWeight = FontWeight.Bold, fontSize = 12.sp),
```
Comment (DAFERMOH) [inline app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt:138]: Again, do not create a new screen just call the one that you created in CreateActivityScreen. And I can't see the part where you assert that the location field is displayed.

File: app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt [hunk 24]
```diff
@@ -1,99 +1,197 @@
...
+    composeTestRule.onNodeWithTag("removeAttendeeButton").performClick()
+    assert(attendees.isEmpty())
+  }
+
+  @Test
+  fun parseFraction_returnsCorrectValueForValidFraction() {
+    val result = parseFraction("5/10", 0)
+    assertEquals(5, result)
+  }
+
```
Comment (DAFERMOH) [inline app/src/androidTest/java/com/android/sample/ui/activity/CreateActivityScreenTest.kt:173]: good testing of the edge cases of the parse function

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt [hunk 25]
```diff
@@ -35,32 +39,44 @@ import androidx.compose.ui.Alignment
...
-    modifier: Modifier = Modifier
 ) {
+  var showDialog by remember { mutableStateOf(false) }
   val activity = listActivityViewModel.selectedActivity.collectAsState().value
-  var title by remember { mutableStateOf(activity?.title) }
-  var description by remember { mutableStateOf(activity?.description) }
-  var creator by remember { mutableStateOf(activity?.creator) }
-  var location by remember { mutableStateOf(activity?.location) }
-  var price by remember { mutableStateOf(activity?.price.toString()) }
-  var placesLeft by remember { mutableStateOf(activity?.placesLeft.toString()) }
```
Comment (DAFERMOH) [inline app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:63]: good to add checks for null cases

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt [hunk 26]
```diff
@@ -180,15 +263,17 @@ fun EditActivityScreen(
...
+                            participants = attendees)
                     listActivityViewModel.updateActivity(updatedActivity)
                     navigationActions.navigateTo(Screen.OVERVIEW)
                   } catch (_: Exception) {}
                 }
               },
-              modifier = Modifier.width(300.dp).height(40.dp).align(Alignment.CenterHorizontally)
-                  .testTag("editButton"),
+              modifier =
+                  Modifier.width(300.dp)
```
Comment (DAFERMOH) [inline app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:273]: LGTM!

File: app/src/main/java/com/android/sample/ui/activity/EditActivity.kt [hunk 27]
```diff
@@ -214,8 +299,11 @@ fun EditActivityScreen(
                 listActivityViewModel.deleteActivityById(activity?.uid ?: "")
                 navigationActions.navigateTo(Screen.OVERVIEW)
               },
-              modifier = Modifier.width(300.dp).height(40.dp).align(Alignment.CenterHorizontally).
-                  testTag("deleteButton"),
```
Comment (DAFERMOH) [inline app/src/main/java/com/android/sample/ui/activity/EditActivity.kt:218]: LGTM!

File: app/src/main/java/com/android/sample/ui/dialogs/AddUserDialog.kt [hunk 28]
```diff
@@ -25,38 +34,59 @@ fun AddUserDialog(
...
-                onAddUser(SimpleUser(name, surname, age.toInt()))
+          OutlinedTextField(
+              modifier = Modifier.testTag("surnameTextFieldUser"),
+              value = surname,
+              onValueChange = { surname = it },
+              label = { Text("Surname") },
+          )
+          OutlinedTextField(
+              modifier = Modifier.testTag("ageTextFieldUser"),
+              value = age,
```
Comment (DAFERMOH) [inline app/src/main/java/com/android/sample/ui/dialogs/AddUserDialog.kt:71]: LGTM!

File: app/src/test/java/com/android/sample/model/activities/ListActivitiesViewModelTest.kt [hunk 29]
```diff
@@ -19,66 +19,127 @@ import org.mockito.kotlin.eq
 
 class ListActivitiesViewModelTest {
```
Comment (DAFERMOH) [inline app/src/test/java/com/android/sample/model/activities/ListActivitiesViewModelTest.kt:21]: same tests

**Review by DAFERMOH — COMMENTED** [review]
LGTM in term of functionnality even though the image inclusion part is not finished yet. Could improve certain points in testing.

**Comment by aminebenji04** [comment]
All the new
